### PR TITLE
[cudagraph] Reuse graph-owned parameter gradients

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #4662
* #4661
* #4660
* __->__ #4659
* #4658

----

- Clear parameter gradients to None in Trainer and FaultTolerantTrainer.
- Keep capture-created gradient buffers alive and restore them after replay.
- Require capture and replay to start without pending gradients.

### CODE REVIEW

File: torchtitan/distributed/cuda_graph.py [hunk 1]
```diff
@@ -227,6 +272,7 @@ def __init__(
         tensor_input_indices: Sequence[int] | None = None,
         *,
         num_warmup_iterations: int,
+        gradient_state: CUDAGraphGradientState | None = None,
```
Comment (tianyu-l) [inline torchtitan/distributed/cuda_graph.py]: would it ever be None?

File: torchtitan/distributed/cuda_graph.py [hunk 2]
```diff
@@ -117,6 +117,50 @@ def unflatten(self, flat_leaves: Sequence[Any]) -> Any:
...
+
+    def __init__(self, parameters: Iterable[torch.nn.Parameter]) -> None:
+        unique_parameters: list[torch.nn.Parameter] = []
+        seen: set[int] = set()
+        for parameter in parameters:
+            if parameter.requires_grad and id(parameter) not in seen:
+                seen.add(id(parameter))
+                unique_parameters.append(parameter)
+        self.parameters = tuple(unique_parameters)
+        self._gradients: tuple[
```
Comment (tianyu-l) [inline torchtitan/distributed/cuda_graph.py]: naming seems not consistent

File: torchtitan/distributed/cuda_graph.py (hunk 1 above)
Comment (fegin) [reply torchtitan/distributed/cuda_graph.py]: yes, this wrapper is also used by optimizer which doesn't have the gradient_state.
